File: kvx_flutter/lib/domain/entities/schedule.dart
import 'package:flutter/material.dart';

enum ScheduleType {
  daily('Hằng ngày'),
  weekly('Hằng tuần'),
  once('Một lần');

  final String displayName;
  const ScheduleType(this.displayName);
}

class Schedule {
  final String id;
  final String deviceId;
  final DateTime time;
  final ScheduleType type;
  final bool isActive;

  const Schedule({
    required this.id,
    required this.deviceId,
    required this.time,
    required this.type,
    this.isActive = true,
  });

  Schedule copyWith({
    String? id,
    String? deviceId,
    DateTime? time,
    ScheduleType? type,
    bool? isActive,
  }) {
    return Schedule(
      id: id ?? this.id,
      deviceId: deviceId ?? this.deviceId,
      time: time ?? this.time,
      type: type ?? this.type,
      isActive: isActive ?? this.isActive,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Schedule && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;
}

File: kvx/Models/Schedule.swift
//
//  Schedule.swift
//  kvx
//
//  Created by Vinh Nguyen on 4/9/26.
//

import Foundation

struct Schedule: Identifiable, Codable {
    let id: String
    let deviceId: String
    let time: Date
    let type: ScheduleType
    var isActive: Bool

    init(id: String = UUID().uuidString, deviceId: String, time: Date, type: ScheduleType, isActive: Bool = true) {
        self.id = id
        self.deviceId = deviceId
        self.time = time
        self.type = type
        self.isActive = isActive
    }

    enum ScheduleType: String, CaseIterable, Codable {
        case daily = "Hằng ngày"
        case weekly = "Hằng tuần"
        case once = "Một lần"
    }

    var timeString: String {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return formatter.string(from: time)
    }

    func copyWith(id: String? = nil, deviceId: String? = nil, time: Date? = nil, type: ScheduleType? = nil, isActive: Bool? = nil) -> Schedule {
        Schedule(
            id: id ?? self.id,
            deviceId: deviceId ?? self.deviceId,
            time: time ?? self.time,
            type: type ?? self.type,
            isActive: isActive ?? self.isActive
        )
    }
}

extension Schedule: Equatable {
    static func == (lhs: Schedule, rhs: Schedule) -> Bool {
        lhs.id == rhs.id
    }
}

extension Schedule: Hashable {
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}
